superset-frontend/packages/superset-ui-core/src/query/extractTimeComparisonFilters.ts

```diff
@@ -0,0 +1,132 @@
+  ensureIsArray,
+  QueryObject,
+  QueryObjectFilterClause,
+} from '@superset-ui/core';
+
+const separator = ' : ';
+const start = 'today';
+const end = 'today';
+const previousCalendarMap = {
+  'previous calendar week':
+    "DATETRUNC(DATEADD(DATETIME('today'), -1, WEEK), WEEK) : DATETRUNC(DATETIME('today'), WEEK)",
+  'previous calendar month':
+    "DATETRUNC(DATEADD(DATETIME('today'), -1, MONTH), MONTH) : DATETRUNC(DATETIME('today'), MONTH)",
+  'previous calendar year':
+    "DATETRUNC(DATEADD(DATETIME('today'), -1, YEAR), YEAR) : DATETRUNC(DATETIME('today'), YEAR)",
+};
+const timeRangeLookup: [RegExp, any][] = [
+  [
+    /^last\s+(day|week|month|quarter|year)$/i,
+    (text: string, unit: string) =>
+      `DATEADD(DATETIME('${start}'), -1, ${unit})`,
+  ],
+  [
+    /^last\s+([0-9]+)\s+(second|minute|hour|day|week|month|year)s?$/i,
+    (text: string, delta: string, unit: string) =>
+      `DATEADD(DATETIME('${start}'), -${delta}, ${unit})`,
+  ],
+  [
+    /^next\s+([0-9]+)\s+(second|minute|hour|day|week|month|year)s?$/i,
+    (text: string, delta: string, unit: string) =>
+      `DATEADD(DATETIME('${end}'), ${delta}, ${unit})`,
+  ],
+  [
+    /^(DATETIME.*|DATEADD.*|DATETRUNC.*|LASTDAY.*|HOLIDAY.*)$/i,
+    (text: string) => text,
+  ],
+  // any plain text should match the following case
+  [/^.*$/i, (text: string) => `DATETIME('${text}')`],
+];
+
+export const getSinceAndUntil = (
+  timeRange: string,
+  start = 'today',
+  end = 'today',
+) => {
+  // eslint-disable-next-line no-underscore-dangle
+  let _timeRange = timeRange;
+  if (timeRange.startsWith('Last') && !timeRange.includes(separator)) {
+    _timeRange = timeRange + separator + end;
+  }
+  if (timeRange.startsWith('Next') && !timeRange.includes(separator)) {
+    _timeRange = start + separator + timeRange;
+  }
+  if (timeRange.startsWith('previous') && !timeRange.includes(separator)) {
+    _timeRange = previousCalendarMap[timeRange];
+  }
+  if (_timeRange?.includes(separator)) {
+    const [since, until] = _timeRange
+      .split(separator)
+      .map(token => token.trim())
+      .map(token => {
+        const matchedPattern = timeRangeLookup.find(
+          item => token.search(item[0]) > -1,
+        ) as [RegExp, any];
+        const args = ensureIsArray(token.match(matchedPattern[0]));
+        return matchedPattern[1](...args);
+      });
+    return [since, until];
+  }
+  return [];
+};
+
+export const getShiftedTimeRange = (
+  timeRange: string,
+  timeShiftType: ComparisonTimeRangeType,
+  start = 'today',
+  end = 'today',
+) => {
+  const [since, until] = getSinceAndUntil(timeRange, start, end);
+  if (!since && !until) {
+    return timeRange;
+  }
+  let unit = 'DAY';
+  let delta = '-1';
+  if (timeShiftType === ComparisonTimeRangeType.Year) {
+    unit = 'YEAR';
+  }
+  if (timeShiftType === ComparisonTimeRangeType.Month) {
+    unit = 'MONTH';
+  }
+  if (timeShiftType === ComparisonTimeRangeType.Week) {
+    unit = 'WEEK';
```

Comment:
Hi @zhaoyongjie. A few of us took a look at this today and we're thinking that we can simplify this logic by using the existing "1 year ago", "1 month ago" and "1 week ago" human readable strings instead of using this enum. And possibly a way to calculate inherit and custom as well. We're going to do some work on Bar and Line which will help us get more familiar with the existing patterns and then we'll come back and update this. Does that work for you? cc @Antonio-RiveroMartnez

Reply:
I thought I would give you an early heads up in case it impacted whether or not you needed as many changes here and consequently if you needed to write a SIP.

Reply:
I think you need to read this function though and check how does time filter work before comments.

Reply:
The goal of this function(`getShiftedTimeRange`) is that generate a shifted time filter base on the previous timeFilter, and implemented in Frontend rather than backend.

Reply:
Hi, yeah, just saying that if you're trying to find a way around the extra code, we'll try to clean this up. If you still want to move the rest of the existing logic to the frontend, then yeah it seems like a SIP would still be worthwhile.